### PR TITLE
Ethernet merge

### PR DESCRIPTION
**Pull request by hargoniX**
This PR adds stm32h7-ethernet into the HAL behind the ethernet feature flag, trying to introduce as little diff as possible in the process of doing so.

Closes: #102

### CODE REVIEW

File: memory.x [hunk 1]
```diff
@@ -45,3 +45,25 @@ _stack_start = ORIGIN(RAM) + LENGTH(RAM);
...
+    } > SRAM1
+  .sram2 (NOLOAD) : ALIGN(4) {
+    *(.sram2 .sram2.*);
+    . = ALIGN(4);
+    } > SRAM2
+  .sram3 (NOLOAD) : ALIGN(4) {
+    *(.sram3 .sram3.*);
+    . = ALIGN(4);
+    } > SRAM3
+} INSERT AFTER .bss;
```
Comment (richardeoin) [inline memory.x:69]: These are needed for any examples involving DMA, but I think we should be more selective about which ones we allow to be used. My suggestion would be:

* Add a comment explaining that these are used for examples
* Remove `.itcm`
* Replace `.sram1` and `.sram2` with a comment explaining that they are commonly used for the CM4 stack/heap, so should not be used in examples
* Add `.sram4`

That gives one RAM on each of the bus matrixes, which should be fine.

Comment (richardeoin) [reply memory.x:69]: Also, adding these invalidate the line L23 `/* Section definitions below this line are not used by stm32h7xx-hal */`

**Comment by ryan-summers** [comment]
@hargoniX / @richardeoin After getting this in, I would like to hear your potential thoughts on the following:
1. Should we pull the `phy` support out (or provide a generic trait such that users can provide their own phy support if required)?
2. Let's clean up the pins definition - we should be able to take a `Pins` structure in the constructor similar to other peripherals
3. I'd really like to clean up the `unsafe`-ness here, but I'm not familiar as to why its originally there.

Curious to hear your thoughts here - we can break this out into an issue if necessary.

**Comment by richardeoin** [comment]
1. I agree it would be nice to pull out the PHY support. The `PHY` trait is there already in ethernet/mod.rs, but currently we use the feature flags to choose an implementation, rather than allowing the user to pass generic object that implements `PHY`. Definitely fixable.

2. Good point, I'd missed that we don't offer a way to check the pins. Definitely we should keep the same `new` /`new_unchecked` pattern.

3. I think there's always going to be some unsafe because of the DMA interaction. But we can definitely reduce it, especially in the public API. A good place to start would be [`ethernet_init`](https://docs.rs/stm32h7-ethernet/0.2.0/stm32h7_ethernet/fn.ethernet_init.html), I don't think it actually needs to be unsafe. But I think users would need unsafe to take the `DesRing` from static storage?

**Comment by richardeoin** [comment]
@ryan-summers Do you want to merge this now, and then continue with those 3 points in another PR?

**Comment by ryan-summers** [comment]
Let's merge as-is and open issues for updates

**Comment by richardeoin** [comment]
Ok!

bors r+

**Pull request merged**